op/model.py

```diff
@@ -312,4 +339,13 @@ def config_get(self):
         return self._run('config-get')
 
     def is_leader(self):
-        return self._run('is-leader')
+        """Obtain the current leadership status for the unit the charm code is executing on.
+
+        The value is cached for the duration of a lease which is 30s in Juju.
+        """
+        if time.monotonic() - self._leader_check_time > self.LEASE_REFRESH_PERIOD or self._is_leader is None:
+            # Record the time before executing is-leader to account for the execution delay and so make the next check to happen earlier.
+            self._leader_check_time = time.monotonic()
+            self._is_leader = self._run('is-leader')
+
+        return self._is_leader
```

Comment:
This change must be on its own PR as it's unrelated to the implementation of application relation data and it's a relevant delta in behavior.  This is way too large by itself already.

Reply:
#59